man/man8/rtmon.8
.TH RTMON 8
.SH "NAME"
rtmon \- listens to and monitors RTnetlink
.SH "SYNOPSIS"
.sp
.ad l
.in +8
.ti -8
.B "rtmon"
.RI "[ " OPTIONS " ] "
.BI "file " FILE
.BR "[ " all
.RI "| " OBJECTS
.RB "]"

.ti -8
.IR OPTIONS
.RI ":= { f[amily] { inet | inet6 | link | help } |"
.RI "-4 | -6 | -0 | -V[ersion] }"

.ti -8
.I OBJECTS
.B ":= [" link "]" "[" address "]" "[" route "]"

.SH DESCRIPTION
This manual page documents briefly the
.B rtmon
command.
.PP
.B rtmon
listens on
.I netlink
socket and monitors routing table changes.

.I rtmon
can be started before the first network configuration command is issued.
For example if you insert:

.B rtmon file /var/log/rtmon.log

in a startup script, you will be able to view the full history later.
Certainly, it is possible to start rtmon at any time. It prepends the history with the state snapshot dumped at the moment of starting.

.SH OPTIONS
.I rtmon supports the following options:
.TP
.B \-Version
Print version and exit.
.TP
.B help
Show summary of options.
.TP
.B file FILE [ all | OBJECTS ]
Log output to FILE. OBJECTS is the list of object types that we
want to monitor. It may contain 'link', 'address', 'route'
and 'all'. 'link' specifies the network device, 'address' the protocol
(IP or IPv6) address on a device, 'route' the routing table entry
and 'all' does what the name says.
.TP
.B \-family [ inet | inet6 | link | help ]
Specify protocol family. 'inet' is IPv4, 'inet6' is IPv6, 'link'
means that no networking protocol is involved and 'help' prints usage information.
.TP
.B \-4
Use IPv4. Shortcut for -family inet.
.TP
.B \-6
Use IPv6. Shortcut for -family inet6.
.TP
.B \-0
Use a special family identifier meaning that no networking protocol is involved. Shortcut for -family link.
.SH USAGE EXAMPLES
.TP
.B # rtmon file /var/log/rtmon.log
Log to file /var/log/rtmon.log, then run:
.TP
.B # ip monitor file /var/log/rtmon.log
to display logged output from file.
.SH SEE ALSO
.BR ip (8)
.BR ip-monitor (8)
.SH AUTHOR
.B rtmon
was written by Alexey Kuznetsov <kuznet@ms2.inr.ac.ru>.
.PP
This manual page was written by Michael Prokop <mika@grml.org>,
for the Debian project (but may be used by others).


ChangeLog
2006-03-21  James Lentini <jlentini@netapp.com>
	
	* Increase size of hw address allowed for ip neigh to allow
	  for IB.

2006-03-14  Russell Stuart <russell-lartc@stuart.id.au>
	
	* Fix missing memset in tc sample
	* Fixes for tc hash samples
	* Add sample divisor

2006-03-10  Alpt <alpt@freaknet.org>

	* Add more rt_proto values

2006-03-10 Dale Sedivec <darkness@caliginous.net>
	
	* Warn when using "handle" instead of "classid" with "tc class"

2006-03-10  Jean Tourrilhes <jt@hpl.hp.com>

	* Fix endless loop in netlink error handling


2006-03-10  Stephen Hemminger  <shemminger@osdl.org>

	* Change default lnstat count to 1
	* Update to 2.6.16 headers
	* Add fake version of include/linux/socket.h to fix warnings

2006-01-12  Patrick McHardy <kaber@trash.net>

	* Handle DCCP in ipxfrm.c to allow using port numbers in the selector.

2006-01-10  Masahide NAKAMURA <nakam@linux-ipv6.org>

	* Add ip link ntable

2006-01-10  Stephen Hemminger  <shemminger@osdl.org>

	* Update headers to santized kernel 2.6.15
	* Fix ipv6 priority option in u32 
	
2006-01-03 Alpt <alpt@freaknet.org>

	* Ip man page addition

2006-01-03  Jamal Hadi Salim <hadi@znyx.com>

	* Documentation for ifb

2005-12-09  Stephen Hemminger  <shemminger@osdl.org>

	* Add corrupt feature to netem

2005-12-02  Stephen Hemminger  <shemminger@osdl.org>

	* Backout ambigious ip command matches

2005-11-22  Stephen Hemminger  <shemminger@osdl.org>

	* Handle ambigious ip command matches

2005-11-22  Patrick McHardy <kaber@trash.net>

	* Add back ip command aliases

2005-11-07  Masahide NAKAMURA <nakam@linux-ipv6.org>

	* Updating for 2.6.14
	- Show UPD{SA,POLICY} message information from kernel instead of error
	- Add lengh check of deleting message from kernel
	- Use macro for struct xfrm_user{sa,policy}_id

	* Minor fix:
	- Add fflush at the end of normal dump

2005-11-01  Jamal Hadi Salim <hadi@znyx.com>

	* Fix handling of XFRM monitor and state

2005-11-01  Stephen Hemminger  <shemminger@osdl.org

	* Update to 2.6.14 sanitized headers

2005-10-24  Patrick McHardy <kaber@trash.net>

	* Fix ip commnad shortcuts

2005-10-12  Stephen Hemminger  <shemminger@osdl.org>

	* Add more CBQ examples from Fedora Core
	* Fix buffer overrun in iproute because of bits vs. bytes confusion
	
2005-10-12  Jamal Hadi Salim <hadi@znyx.com>

	* Fix ip rule flush, need to reopen rtnl

2005-10-07  Stephen Hemminger  <shemminger@osdl.org>

	* Reenable ip mroute

2005-10-07  Mike Frysinger <vapier@gentoo.org>

	* Handle pfifo_fast that has no qopt without segfaulting

2005-10-05  Mads Martin Joergensen <mmj@suse.de>

	* Trivial netem ccopts

2005-10-04  Jerome Borsboom <j.borsboom@erasmusmc.nl>

	* Fix regression in ip addr (libnetlink) handling

2005-09-21  Stephen Hemminger  <shemminger@osdl.org>

	* Fix uninitialized memory and leaks with valgrind
	  Reported by Redhat

2005-09-01   Mike Frysinger <vapier@gentoo.org>

	* Fix build issues with netem tables (parallel make and HOSTCC)
	
2005-09-01  Stephen Hemminger  <shemminger@osdl.org>

	* Integrate support for DCCP into 'ss' (from acme)
	* Add -batch option to ip.
	* Update to 2.6.14 headers

2005-09-01  Eric Dumazet <dada1@cosmosbay.com>
	
	* Fix lnstat : First column should not be summed

2005-08-16  Stephen Hemminger  <shemminger@osdl.org>

	* Limit ip route flush to 10 rounds.
	* Cleanup ip rule flush error message
	
2005-08-08  Stephen Hemminger  <shemminger@osdl.org>

	* Update to 2.6.13+ kernel headers
	* Fix array overrun in paretonormal
	* Fix ematch to not include dropped fields from skb.
	
2005-07-14  Thomas Graf <tgraf@suug.ch>

	* Make ematch bison/lex build with common flex
	
2005-07-10  Stephen Hemminger  <shemminger@osdl.org>
	
	* Fix Gcc 4.0 build warnings signed/unsigned

2005-06-23  Jamal Hadi Salim <hadi@znyx.com>

	* Fix for options process with ipt

2005-06-23  Thomas Graf <tgraf@suug.ch>
	
	* Add extended matches (nbyte, cmp, u32, meta)
	* Add basic classifier
	* Fix clean/distclean makefile targets
	* update local header file copies
	* IPv4 multipath algorithm selection support
	* cscope Makefile target
	* Fix off-by-one while generating argument vector
	  in batched mode.
	* Assume stdin if no argument is given to -batch

2005-06-22  Stephen Hemminger  <shemminger@osdl.org>

	* Update include files to 2.6.12
	* Add ss support for TCP_CONG

2005-06-13  Steven Whitehouse <steve@chygwyn.com>

	* Decnet doc's update

2005-06-07  Stephen Hemminger  <shemminger@osdl.org>

	* Fix 'ip link' map to handle case where device gets autoloaded
	  by using if_nametoindex as fallback
	* Device indices are unsigned not int.

2005-06-07   Masahide NAKAMURA <nakam@linux-ipv6.org>
	
	* [ip] show timestamp when using '-t' option.
	* [ip] remove duplicated code for expired message of xfrm.
	* [ip] add "deleteall" command for xfrm;
	  "flush" uses kernel's flush interface and
	  "deleteall" uses legacy iproute2's flush feature like
	   getting-and-deleting-for-each.

2005-03-30  Stephen Hemminger  <shemminger@osdl.org>

	* include/linux/netfilter_ipv4/ip_tables.h dont include compiler.h
	  because it isn't needed and not on all systems
	* Update rtnetlink.h and pkt_cls.h to be stripped versions
	  of headers from 2.6.12-rc1

2005-03-30  Jamal Hadi Salim <hadi@znyx.com>

	* Proper verison of iptables headers (from 1.3.1)
	* Set revision file in m_ipt
	* Fix action_util naming in mirred
	* don't call ll_init_map in mirred

2005-03-19  Thomas Graf <tgraf@suug.ch>

	* Warn about wildcard deletions and provide IFA_ADDRESS upon
	  deletions to enforce prefix length validation for IPv4.
	* Fix netlink message alignment when the last routing attribute added
	  has a data length not aligned to RTA_ALIGNTO.
	
2005-03-30  Masahide NAKAMURA <nakam@linux-ipv6.org>
	
	* ipv6 xfrm allocspi and monitor support.
	
2005-03-29  Stephen Hemminger  <shemminger@osdl.org>

	* switch to stack for netem tables

2005-03-18  Stephen Hemminger  <shemminger@osdl.org>

	* add -force option to batch mode

ip/rtmon.c
/* SPDX-License-Identifier: GPL-2.0-or-later */
/*
 * rtmon.c		RTnetlink listener.
 *
 * Authors:	Alexey Kuznetsov, <kuznet@ms2.inr.ac.ru>
 */

#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <fcntl.h>
#include <sys/socket.h>
#include <sys/time.h>
#include <netinet/in.h>
#include <string.h>

#include "version.h"

#include "utils.h"
#include "libnetlink.h"

static int init_phase = 1;

static void write_stamp(FILE *fp)
{
	char buf[128];
	struct nlmsghdr *n1 = (void *)buf;
	struct timeval tv;

	n1->nlmsg_type = NLMSG_TSTAMP;
	n1->nlmsg_flags = 0;
	n1->nlmsg_seq = 0;
	n1->nlmsg_pid = 0;
	n1->nlmsg_len = NLMSG_LENGTH(4*2);
	gettimeofday(&tv, NULL);
	((__u32 *)NLMSG_DATA(n1))[0] = tv.tv_sec;
	((__u32 *)NLMSG_DATA(n1))[1] = tv.tv_usec;
	fwrite((void *)n1, 1, NLMSG_ALIGN(n1->nlmsg_len), fp);
}

static int dump_msg(struct rtnl_ctrl_data *ctrl,
		    struct nlmsghdr *n, void *arg)
{
	FILE *fp = (FILE *)arg;

	if (!init_phase)
		write_stamp(fp);
	fwrite((void *)n, 1, NLMSG_ALIGN(n->nlmsg_len), fp);
	fflush(fp);
	return 0;
}

static int dump_msg2(struct nlmsghdr *n, void *arg)
{
	return dump_msg(NULL, n, arg);
}

static void usage(void)
{
	fprintf(stderr,
		"Usage: rtmon [ OPTIONS ] file FILE [ all | OBJECTS ]\n"
		"OPTIONS := { -f[amily] { inet | inet6 | link | help } |\n"
		"             -4 | -6 | -0 | -V[ersion] }\n"
		"OBJECTS := [ link ] [ address ] [ route ]\n");
	exit(-1);
}

int
main(int argc, char **argv)
{
	FILE *fp;
	struct rtnl_handle rth;
	int family = AF_UNSPEC;
	unsigned int groups = ~0U;
	int llink = 0;
	int laddr = 0;
	int lroute = 0;
	char *file = NULL;

	while (argc > 1) {
		if (matches(argv[1], "-family") == 0) {
			argc--;
			argv++;
			if (argc <= 1)
				missarg("family type");
			if (strcmp(argv[1], "inet") == 0)
				family = AF_INET;
			else if (strcmp(argv[1], "inet6") == 0)
				family = AF_INET6;
			else if (strcmp(argv[1], "link") == 0)
				family = AF_INET6;
			else if (strcmp(argv[1], "help") == 0)
				usage();
			else {
				fprintf(stderr, "Protocol ID \"%s\" is unknown, try \"rtmon help\".\n", argv[1]);
				exit(-1);
			}
		} else if (strcmp(argv[1], "-4") == 0) {
			family = AF_INET;
		} else if (strcmp(argv[1], "-6") == 0) {
			family = AF_INET6;
		} else if (strcmp(argv[1], "-0") == 0) {
			family = AF_PACKET;
		} else if (matches(argv[1], "-Version") == 0) {
			printf("rtmon utility, iproute2-%s\n", version);
			exit(0);
		} else if (matches(argv[1], "file") == 0) {
			argc--;
			argv++;
			if (argc <= 1)
				missarg("file");
			file = argv[1];
		} else if (matches(argv[1], "link") == 0) {
			llink = 1;
			groups = 0;
		} else if (matches(argv[1], "address") == 0) {
			laddr = 1;
			groups = 0;
		} else if (matches(argv[1], "route") == 0) {
			lroute = 1;
			groups = 0;
		} else if (strcmp(argv[1], "all") == 0) {
			groups = ~0U;
		} else if (matches(argv[1], "help") == 0) {
			usage();
		} else {
			fprintf(stderr, "Argument \"%s\" is unknown, try \"rtmon help\".\n", argv[1]);
			exit(-1);
		}
		argc--;	argv++;
	}

	if (file == NULL) {
		fprintf(stderr, "Not enough information: argument \"file\" is required\n");
		exit(-1);
	}
	if (llink)
		groups |= nl_mgrp(RTNLGRP_LINK);
	if (laddr) {
		if (!family || family == AF_INET)
			groups |= nl_mgrp(RTNLGRP_IPV4_IFADDR);
		if (!family || family == AF_INET6)
			groups |= nl_mgrp(RTNLGRP_IPV6_IFADDR);
	}
	if (lroute) {
		if (!family || family == AF_INET)
			groups |= nl_mgrp(RTNLGRP_IPV4_ROUTE);
		if (!family || family == AF_INET6)
			groups |= nl_mgrp(RTNLGRP_IPV6_ROUTE);
	}

	fp = fopen(file, "w");
	if (fp == NULL) {
		perror("Cannot fopen");
		exit(-1);
	}

	if (rtnl_open(&rth, groups) < 0)
		exit(1);

	if (rtnl_linkdump_req(&rth, AF_UNSPEC) < 0) {
		perror("Cannot send dump request");
		exit(1);
	}

	write_stamp(fp);

	if (rtnl_dump_filter(&rth, dump_msg2, fp) < 0) {
		fprintf(stderr, "Dump terminated\n");
		return 1;
	}

	init_phase = 0;

	if (rtnl_listen(&rth, dump_msg, (void *)fp) < 0)
		exit(2);

	exit(0);
}
